README.md
# ember-ajax

[![npm version](https://badge.fury.io/js/ember-ajax.svg)](https://badge.fury.io/js/ember-ajax)
[![Travis CI Build Status](https://travis-ci.org/ember-cli/ember-ajax.svg?branch=master)](https://travis-ci.org/ember-cli/ember-ajax)
[![Ember Observer Score](http://emberobserver.com/badges/ember-ajax.svg)](http://emberobserver.com/addons/ember-ajax)

Service for making AJAX requests in Ember applications.

- customizable service
- returns RSVP promises
- improved error handling
- ability to specify request headers

#### Contents 

- [Getting Started](#Getting-started)
- [Ajax Service](#Ajax-Service)
- [Usage with Ember Data](#Usage-with-Ember-Data)
- [Stand-Alone Usage](#Stand-Alone-Usage)
- [CONTRIBUTING](.github/CONTRIBUTING.md)

## Getting started

If you're just starting out, you already have `ember-ajax` installed! However, if it's missing from your `package.json`, you can add it by doing:

```sh
ember install ember-ajax
```

To use the ajax service, inject the `ajax` service into your route or component.

```js
import Ember from 'ember';

export default Ember.Route.extend({
  ajax: Ember.inject.service(),
  model() {
    return this.get('ajax').request('/posts');
  }
});
```

## Ajax Service

### Basic Usage

The AJAX service provides methods to be used to make AJAX requests, similar to
the way that you would use `jQuery.ajax`. In fact, `ember-ajax` is a wrapper
around jQuery's method, and can be configured in much the same way.

In general, you will use the `request(url, options)` method, where `url` is the
destination of the request and `options` is a configuration hash for
[`jQuery.ajax`](http://api.jquery.com/jQuery.ajax/#jQuery-ajax-settings).

```javascript
import Ember from 'ember';

export default Ember.Controller.extend({
  ajax: Ember.inject.service(),
  actions: {
    sendRequest() {
      return this.get('ajax').request('/posts', {
        method: 'POST',
        data: {
          foo: 'bar'
        }
      });
    }
  }
});
```

In this example, `this.get('ajax').request()` will return a promise with the
result of the request. Your handler code inside `.then` or `.catch` will
automatically be wrapped in an Ember run loop for maximum compatibility with
Ember, right out of the box.

### HTTP-verbed methods

You can skip setting the `method` or `type` keys in your `options` object when
calling `request(url, options)` by instead calling `post(url, options)`,
`put(url, options)`, `patch(url, options)` or `del(url, options)`.

```js
post('/posts', { data: { title: 'Ember' } }); // Makes a POST request to /posts
put('/posts/1', { data: { title: 'Ember' } }); // Makes a PUT request to /posts/1
patch('/posts/1', { data: { title: 'Ember' } }); // Makes a PATCH request to /posts/1
del('/posts/1'); // Makes a DELETE request to /posts/1
```

### Custom Request Headers

`ember-ajax` allows you to specify headers to be used with a request. This is
especially helpful when you have a session service that provides an auth token
that you have to include with the requests to authorize your requests.

To include custom headers to be used with your requests, you can specify
`headers` hash on the `Ajax Service`.

```js
// app/services/ajax.js

import Ember from 'ember';
import AjaxService from 'ember-ajax/services/ajax';

export default AjaxService.extend({
  session: Ember.inject.service(),
  headers: Ember.computed('session.authToken', {
    get() {
      let headers = {};
      const authToken = this.get('session.authToken');
      if (authToken) {
        headers['auth-token'] = authToken;
      }
      return headers;
    }
  })
});
```

Headers by default are only passed if the hosts match, or the request is a relative path.
You can overwrite this behavior by either passing a host in with the request, setting the
host for the ajax service, or by setting an array of `trustedHosts` that can be either
an array of strings or regexes.

```js
// app/services/ajax.js

import Ember from 'ember';
import AjaxService from 'ember-ajax/services/ajax';

export default AjaxService.extend({
  trustedHosts: [/\.example\./, 'foo.bar.com']
});
```

### Custom Endpoint Path

The `namespace` property can be used to prefix requests with a specific url namespace.

```js
// app/services/ajax.js
import Ember from 'ember';
import AjaxService from 'ember-ajax/services/ajax';

export default AjaxService.extend({
  namespace: '/api/v1'
});
```

`request('/users/me')` would now target `/api/v1/users/me`

If you need to override the namespace for a custom request, use the `namespace` as an option to the request methods.

```js
// GET /api/legacy/users/me
request('/users/me', { namespace: '/api/legacy' });
```

### Custom Host

`ember-ajax` allows you to specify a host to be used with a request. This is
especially helpful so you don't have to continually pass in the host along
with the path, makes `request()` a bit cleaner.

To include a custom host to be used with your requests, you can specify `host`
property on the `Ajax Service`.

```js
// app/services/ajax.js

import Ember from 'ember';
import AjaxService from 'ember-ajax/services/ajax';

export default AjaxService.extend({
  host: 'http://api.example.com'
});
```

That allows you to only have to make a call to `request()` as such:

```js
// GET http://api.example.com/users/me
request('/users/me');
```

### Custom Content-Type

`ember-ajax` allows you to specify a default [Content-Type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type) header to be used with a request.

To include a custom Content-Type you can specify `contentType` property on the `Ajax Service`.

```js
// app/services/ajax.js

import Ember from 'ember';
import AjaxService from 'ember-ajax/services/ajax';

export default AjaxService.extend({
  contentType: 'application/json; charset=utf-8'
});
```

You can also override the Content-Type per `request` with the `options` parameter.

#### Customize isSuccess

Some APIs respond with status code 200, even though an error has occurred and
provide a status code in the payload. With the service, you can account for this 
behaviour by overwriting the `isSuccess` method.

```js
// app/services/ajax.js

import AjaxService from 'ember-ajax/services/ajax';

export default AjaxService.extend({
  isSuccess(status, headers, payload) {
    let isSuccess = this._super(...arguments);
    if (isSuccess && payload.status) {
      // when status === 200 and payload has status property,
      // check that payload.status is also considered a success request
      return this._super(payload.status);
    }
    return isSuccess;
  }
});
```

### Error handling

`ember-ajax` provides built in error classes that you can use to check the error
that was returned by the response. This allows you to restrict determination of
error result to the service instead of sprinkling it around your code.

#### Built in error types

`ember-ajax` has built-in error types that will be returned from the service in 
the event of an error:

- `BadRequestError` (400)
- `UnauthorizedError`(401)
- `ForbiddenError`(403)
- `NotFoundError` (404)
- `InvalidError`(422)
- `ServerError` (5XX)
- `AbortError`
- `TimeoutError`

All of the above errors are subtypes of `AjaxError`.

#### Error detection helpers

`ember-ajax` comes with helper functions for matching response errors to their 
respective `ember-ajax` error type. Each of the errors listed above has a 
corresponding `is*` function (e.g., `isBadRequestError`).

Use of these functions is **strongly encouraged** to help eliminate the need for 
boilerplate error detection code.

```js
import Ember from 'ember';
import {
  isAjaxError,
  isNotFoundError,
  isForbiddenError
} from 'ember-ajax/errors';

export default Ember.Route.extend({
  ajax: Ember.inject.service(),
  model() {
    const ajax = this.get('ajax');

    return ajax.request('/user/doesnotexist').catch(function(error) {
      if (isNotFoundError(error)) {
        // handle 404 errors here
        return;
      }

      if (isForbiddenError(error)) {
        // handle 403 errors here
        return;
      }

      if (isAjaxError(error)) {
        // handle all other AjaxErrors here
        return;
      }

      // other errors are handled elsewhere
      throw error;
    });
  }
});
```

If your errors aren't standard, the helper function for that error type can be 
used as the base to build your custom detection function.

#### Access the response in case of error

If you need to access the json response of a request that failed, you can use the 
`raw` method instead of `request`.

```js
this.get('ajax')
  .raw(url, options)
  .then(({ response }) => this.handleSuccess(response))
  .catch(({ response, jqXHR, payload }) => this.handleError(response));
```

Note that in this use case there's no access to the error object. You can inspect 
the `jqXHR` object for additional information about the failed request. In 
particular `jqXHR.status` returns the relevant HTTP error code.

## Usage with Ember Data

Ember AJAX provides a mixin that can be used in an Ember Data Adapter to avoid 
the networking code provided by Ember Data and rely on Ember AJAX instead. This 
serves as a first step toward true integration of Ember AJAX into Ember Data.

To use the mixin, you can include the mixin into an Adapter, like so:

```javascript
// app/adapters/application.js
import DS from 'ember-data';
import AjaxServiceSupport from 'ember-ajax/mixins/ajax-support';

export default DS.JSONAPIAdapter.extend(AjaxServiceSupport);
```

That's all the configuration required! If you want to customize the adapter, such 
as using an alternative AJAX service (like one you extended yourself), hooks to 
do so are provided; check out the mixin's implementation for details.

Note that instead of using the Ember Data error checking code in your application, 
you should use the ones provided by Ember AJAX.

## Stand-Alone Usage

If you aren't using Ember Data and do not have access to services, you
can import the ajax utility like so:

```js
import request from 'ember-ajax/request';

export default function someUtility(url) {
  var options = {
    // request options
  };

  return request(url, options).then(response => {
    // `response` is the data from the server
    return response;
  });
}
```

Which will have the same API as the `ajax` service. If you want the raw jQuery 
XHR object then you can use the `raw` method instead:

```js
import raw from 'ember-ajax/raw';

export default function someOtherUtility(url) {
  var options = {
    // raw options
  };

  return raw(url, options).then(result => {
    // `result` is an object containing `response` and `jqXHR`, among other items
    return result;
  });
}
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="5.1.2"></a>
## [5.1.2](https://github.com/ember-cli/ember-ajax/compare/v5.1.1...v5.1.2) (2022-10-11)


### Bug Fixes

* resolve `ember-polyfills.deprecate-assign` deprecation warnings ([5175d40](https://github.com/ember-cli/ember-ajax/commit/5175d40))



<a name="5.0.0"></a>
## [5.0.0](https://github.com/ember-cli/ember-ajax/compare/v4.0.2...v5.0.0) (2019-03-05)


### Breaking Changes

- Update `ember-cli-babel` to v7.x
  (this requires apps to use Ember CLI 2.13 or above)

- Extend the native `Error` class instead of `EmberError`
  (fixes compatibility with Ember 3.8)

- Drop Node 4 from `engines` declaration in `package.json`
  (CI had already been running on Node 6 for a while already)


### Bug Fixes

- Restore ability to override the `ajax` service using a JS file


### Documentation

- Add "Compatibility" section to the `README` file


<a name="4.0.2"></a>
## [4.0.2](https://github.com/ember-cli/ember-ajax/compare/v4.0.1...v4.0.2) (2019-01-07)


### Bug Fixes

* **deprecation:** the new EmberObject is deprecated ([80e6e15](https://github.com/ember-cli/ember-ajax/commit/80e6e15))



<a name="4.0.1"></a>
## [4.0.1](https://github.com/ember-cli/ember-ajax/compare/v4.0.0...v4.0.1) (2018-12-10)



<a name="4.0.0"></a>
# [4.0.0](https://github.com/ember-cli/ember-ajax/compare/v3.1.2...v4.0.0) (2018-10-23)



<a name="3.1.2"></a>
## [3.1.2](https://github.com/ember-cli/ember-ajax/compare/v3.1.1...v3.1.2) (2018-10-09)


### Bug Fixes

* bump typescript to get working d.ts ([975724d](https://github.com/ember-cli/ember-ajax/commit/975724d)), closes [#388](https://github.com/ember-cli/ember-ajax/issues/388)
* is*Error type definitions ([101c357](https://github.com/ember-cli/ember-ajax/commit/101c357))
* skip instead of returning url when namespace is present in url ([fe518dc](https://github.com/ember-cli/ember-ajax/commit/fe518dc))



Changes for each version can be found as part of the [release notes on Github](https://github.com/ember-cli/ember-ajax/releases).


addon/service.js
import Ember from 'ember';

import {
  AjaxError,
  UnauthorizedError,
  InvalidError,
  ForbiddenError
} from './errors';
import parseResponseHeaders from './utils/parse-response-headers';

const {
  get
} = Ember;

/**
  ### Headers customization

  Some APIs require HTTP headers, e.g. to provide an API key. Arbitrary
  headers can be set as key/value pairs on the `RESTAdapter`'s `headers`
  object and Ember Data will send them along with each ajax request.

  ```app/services/ajax
  import AjaxService from 'ember-ajax/service';

  export default AjaxService.extend({
    headers: {
      "API_KEY": "secret key",
      "ANOTHER_HEADER": "Some header value"
    }
  });
  ```

  `headers` can also be used as a computed property to support dynamic
  headers.

  ```app/services/ajax.js
  import Ember from 'ember';
  import AjaxService from 'ember-ajax/service';

  export default AjaxService.extend({
    session: Ember.inject.service(),
    headers: Ember.computed("session.authToken", function() {
      return {
        "API_KEY": this.get("session.authToken"),
        "ANOTHER_HEADER": "Some header value"
      };
    })
  });
  ```

  In some cases, your dynamic headers may require data from some
  object outside of Ember's observer system (for example
  `document.cookie`). You can use the
  [volatile](/api/classes/Ember.ComputedProperty.html#method_volatile)
  function to set the property into a non-cached mode causing the headers to
  be recomputed with every request.

  ```app/services/ajax.js
  import Ember from 'ember';
  import AjaxService from 'ember-ajax/service';

  export default AjaxService.extend({
    session: Ember.inject.service(),
    headers: Ember.computed("session.authToken", function() {
      return {
        "API_KEY": Ember.get(document.cookie.match(/apiKey\=([^;]*)/), "1"),
        "ANOTHER_HEADER": "Some header value"
      };
    }).volatile()
  });
  ```

**/
export default Ember.Service.extend({

  request(url, type, options) {
    const hash = this.options(url, type, options);

    return new Ember.RSVP.Promise((resolve, reject) => {

      hash.success = (payload, textStatus, jqXHR) => {
        let response = this.handleResponse(
          jqXHR.status,
          parseResponseHeaders(jqXHR.getAllResponseHeaders()),
          payload
        );

        if (response instanceof AjaxError) {
          reject(response);
        } else {
          resolve(response);
        }
      };

      hash.error = (jqXHR, textStatus, errorThrown) => {
        let error;

         if (!(error instanceof Error)) {
           if (errorThrown instanceof Error) {
             error = errorThrown;
           } else {
             error = this.handleResponse(
               jqXHR.status,
               parseResponseHeaders(jqXHR.getAllResponseHeaders()),
               this.parseErrorResponse(jqXHR.responseText) || errorThrown
             );
           }
         }
         reject(error);
      };

      Ember.$.ajax(hash);
    }, `ember-ajax: ${hash.type} to ${url}`);
  },

  /**
    @method ajaxOptions
    @private
    @param {String} url
    @param {String} type The request type GET, POST, PUT, DELETE etc.
    @param {Object} options
    @return {Object}
  */
  options(url, type, options) {
    var hash = options || {};
    hash.url = url;
    hash.type = type;
    hash.dataType = 'json';
    hash.context = this;

    if (hash.data && type !== 'GET') {
      hash.contentType = 'application/json; charset=utf-8';
      hash.data = JSON.stringify(hash.data);
    }

    var headers = get(this, 'headers');
    if (headers !== undefined) {
      hash.beforeSend = function (xhr) {
        Object.keys(headers).forEach((key) =>  xhr.setRequestHeader(key, headers[key]));
      };
    }

    return hash;
  },

  /**
   Takes an ajax response, and returns the json payload or an error.

   By default this hook just returns the json payload passed to it.
   You might want to override it in two cases:

   1. Your API might return useful results in the response headers.
   Response headers are passed in as the second argument.

   2. Your API might return errors as successful responses with status code
   200 and an Errors text or object.
   @method handleResponse
   @param  {Number} status
   @param  {Object} headers
   @param  {Object} payload
   @return {Object | DS.AdapterError} response
 */
 handleResponse(status, headers, payload) {
   if (this.isSuccess(status, headers, payload)) {
     return payload;
   } else if (this.isUnauthorized(status, headers, payload)) {
     return new UnauthorizedError(payload.errors);
   } else if (this.isForbidden(status, headers, payload)){
     return new ForbiddenError(payload.errors);
   } else if (this.isInvalid(status, headers, payload)) {
     return new InvalidError(payload.errors);
   }

   let errors = this.normalizeErrorResponse(status, headers, payload);

   return new AjaxError(errors);
 },

  /**
   Default `handleResponse` implementation uses this hook to decide if the
   response is a an authorized error.
   @method isUnauthorized
   @param  {Number} status
   @param  {Object} headers
   @param  {Object} payload
   @return {Boolean}
 */
  isUnauthorized(status/*, headers, payload */) {
    return status === 401;
  },

  /**
     Default `handleResponse` implementation uses this hook to decide if the
     response is a forbidden error.
     @method isForbidden
     @param  {Number} status
     @param  {Object} headers
     @param  {Object} payload
     @return {Boolean}
   */
  isForbidden(status/*, headers, payload */) {
    return status === 403;
  },

  /**
    Default `handleResponse` implementation uses this hook to decide if the
    response is a an invalid error.
    @method isInvalid
    @param  {Number} status
    @param  {Object} headers
    @param  {Object} payload
    @return {Boolean}
  */
  isInvalid(status/*, headers, payload */) {
  return status === 422;
  },

   /**
    Default `handleResponse` implementation uses this hook to decide if the
    response is a success.
    @method isSuccess
    @param  {Number} status
    @param  {Object} headers
    @param  {Object} payload
    @return {Boolean}
  */
  isSuccess(status/*, headers, payload */) {
    return status >= 200 && status < 300 || status === 304;
  },

  /**
    @method parseErrorResponse
    @private
    @param {String} responseText
    @return {Object}
  */
  parseErrorResponse(responseText) {
    var json = responseText;

    try {
      json = Ember.$.parseJSON(responseText);
    } catch (e) {}

    return json;
  },

  /**
    @method normalizeErrorResponse
    @private
    @param  {Number} status
    @param  {Object} headers
    @param  {Object} payload
    @return {Object} errors payload
  */
  normalizeErrorResponse(status, headers, payload) {
    if (payload && typeof payload === 'object' && payload.errors) {
      return payload.errors;
    } else {
      return [
        {
          status: `${status}`,
          title: "The backend responded with an error",
          detail: `${payload}`
        }
      ];
    }
  }
});
